CBlockHeader::GetHash() must serialize the block header

Disclaimer: This is based on code review only, and not testing. I could be wrong.

Bitcoin:

```c
uint256 CBlockHeader::GetHash() const
{
    return SerializeHash(*this);
}
```

Dash:

```c
uint256 CBlockHeader::GetHash() const
{
    return HashX11(BEGIN(nVersion), END(nNonce));
}
```

This omits serialization. I see no serialization inside `HashX11` either. Some header fields are 32-bit integers in host byte order (the `uint256` ones are host-independent, though). Thus, I expect that Dash core tree as it currently is to malfunction on big-endian systems, and only work right on the currently prevalent little-endian systems.

The fix may be something like this (untested):

```c
uint256 CBlockHeader::GetHash() const
{
    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
    ss << *this;
    return HashX11((const char *)&ss[0], (const char *)&ss[0] + ss.size());
}
```

Submitted a PR, but this is still untested on my end, especially as the develop branch no longer builds on the CentOS 7 box with its standard gcc:

```
configure: error: *** A compiler with support for C++14 language features is required.
```

The master branch builds fine (I assume that'll end when you merge develop into master), and I could test the change there, or install a newer gcc or try on another box, but I did not. I'll rely on your testing. ;-)

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -398,9 +398,6 @@ libdash_consensus_a_SOURCES = \
   script/script_error.h \
   serialize.h \
   streams.h \
-  support/allocators/zeroafterfree.h \
-  support/cleanse.cpp \
-  support/cleanse.h \
   tinyformat.h \
   uint256.cpp \
   uint256.h \

--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -397,6 +397,10 @@ libdash_consensus_a_SOURCES = \
   script/script_error.cpp \
   script/script_error.h \
   serialize.h \
+  streams.h \
+  support/allocators/zeroafterfree.h \
+  support/cleanse.cpp \
+  support/cleanse.h \
   tinyformat.h \
   uint256.cpp \
   uint256.h \

--- a/src/primitives/block.cpp
+++ b/src/primitives/block.cpp
@@ -5,17 +5,18 @@
 
 #include "primitives/block.h"
 
-#include "streams.h"
 #include "hash.h"
+#include "streams.h"
 #include "tinyformat.h"
 #include "utilstrencodings.h"
 #include "crypto/common.h"
 
 uint256 CBlockHeader::GetHash() const
 {
-    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    std::vector<unsigned char> vch(80);
+    CVectorWriter ss(SER_NETWORK, PROTOCOL_VERSION, vch, 0);
     ss << *this;
-    return HashX11((const char *)&ss[0], (const char *)&ss[0] + ss.size());
+    return HashX11((const char *)vch.data(), (const char *)vch.data() + vch.size());
 }
 
 std::string CBlock::ToString() const

--- a/src/primitives/block.cpp
+++ b/src/primitives/block.cpp
@@ -5,6 +5,7 @@
 
 #include "primitives/block.h"
 
+#include "streams.h"
 #include "hash.h"
 #include "tinyformat.h"
 #include "utilstrencodings.h"
@@ -12,7 +13,9 @@
 
 uint256 CBlockHeader::GetHash() const
 {
-    return HashX11(BEGIN(nVersion), END(nNonce));
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << *this;
+    return HashX11((const char *)&ss[0], (const char *)&ss[0] + ss.size());
 }
 
 std::string CBlock::ToString() const

